### PR TITLE
feat: improve the integration of the banned and the delayed/retain feature 

### PR DESCRIPTION
**Pull request by lafirest**
Port from v5

### CODE REVIEW

File: apps/emqx_retainer/test/emqx_retainer_SUITE.erl [hunk 1]
```diff
@@ -189,6 +190,46 @@ t_stop_publish_clear_msg(_) ->
...
+        by => <<"test">>,
+        reason => <<"test">>,
+        at => Now,
+        until => Now + 120
+    }),
+
+    timer:sleep(100),
+    snabbkaffe:start_trace(),
+    {ok, #{}, [0]} = emqtt:subscribe(C1, <<"retained/+">>, [{qos, 0}, {rh, 0}]),
+    timer:sleep(500),
```
Comment (savonarola) [inline apps/emqx_retainer/test/emqx_retainer_SUITE.erl:224]: Maybe we could wait for `ignore_retained_message_deliver` to reduce test run time?

File: lib-ce/emqx_modules/test/emqx_mod_delayed_SUITE.erl [hunk 2]
```diff
@@ -78,3 +79,38 @@ t_delayed_message(_) ->
...
+    snabbkaffe:start_trace(),
+    lists:foreach(
+        fun(ClientId) ->
+            Msg = emqx_message:make(ClientId, <<"$delayed/1/bc">>, <<"payload">>),
+            emqx_mod_delayed:on_message_publish(Msg)
+        end,
+        [ClientId1, ClientId1, ClientId1, ClientId2, ClientId2]
+    ),
+
+    timer:sleep(2000),
```
Comment (savonarola) [inline lib-ce/emqx_modules/test/emqx_mod_delayed_SUITE.erl:107]: Here too, maybe we could `wait_async` for `ignore_delayed_message_publish`?

File: changes/v4.4.14-en.md [hunk 3]
```diff
@@ -3,6 +3,12 @@
 ## Enhancements
 - Add a password complexity requirement when adding or modifying Dashboard users via the API. Now passwords must contain at least 2 of alphabetic, numeric and special characters, and must be 8 to 64 characters long [#9696](https://github.com/emqx/emqx/pull/9696).
 
+- Improve the integration of the `banned` and the `delayed` feature [#9790](https://github.com/emqx/emqx/pull/9790).
```
Comment (zmstone) [inline changes/v4.4.14-en.md:6]: 4.4.14 is already released. please move on to 4.4.15

File: apps/emqx_retainer/src/emqx_retainer.erl [hunk 4]
```diff
@@ -332,3 +334,21 @@ refresh_timestamp_expiry(Msg = #message{headers =
 
 refresh_timestamp_expiry(Msg, Now) ->
     Msg#message{timestamp = Now}.
+
+filter_banned_messages(Msgs) ->
```
Comment (zmstone) [inline apps/emqx_retainer/src/emqx_retainer.erl:338]:
```suggestion
drop_banned_messages(Msgs) ->
```

**Review by savonarola — APPROVED** [review]

File: apps/emqx_retainer/src/emqx_retainer.erl [hunk 5]
```diff
@@ -74,11 +75,12 @@ on_session_subscribed(_, Topic, #{rh := Rh, is_new := IsNew}) ->
...
-    Msgs = case emqx_topic:wildcard(Topic) of
-               false -> read_messages(Topic);
-               true  -> match_messages(Topic)
-           end,
+    MsgsT = case emqx_topic:wildcard(Topic) of
+                false -> read_messages(Topic);
+                true  -> match_messages(Topic)
+            end,
     Now = erlang:system_time(millisecond),
+    Msgs = drop_banned_messages(MsgsT),
```
Comment (thalesmg) [inline apps/emqx_retainer/src/emqx_retainer.erl:83]: Q: should we also delete messages from banned clients from storage at this point? :thinking:

**Review by thalesmg — APPROVED** [review]

**Comment by zmstone** [comment]
need a rebase

**Review by zmstone — APPROVED** [review]

**Pull request merged**